File: src/main/java/net/ndrei/teslacorelib/TeslaCoreLib.java
package net.ndrei.teslacorelib;

import net.minecraft.creativetab.CreativeTabs;
import net.minecraft.item.ItemStack;
import net.minecraftforge.fml.common.Mod;
import net.minecraftforge.fml.common.event.FMLPreInitializationEvent;
import net.minecraftforge.fml.common.network.NetworkRegistry;
import net.minecraftforge.fml.relauncher.Side;
import net.ndrei.teslacorelib.capabilities.TeslaCoreCapabilities;
import net.ndrei.teslacorelib.gui.TeslaCoreGuiProxy;
import net.ndrei.teslacorelib.items.*;
import net.ndrei.teslacorelib.netsync.ITeslaCorePackets;
import net.ndrei.teslacorelib.netsync.TeslaCorePackets;
import net.ndrei.teslacorelib.test.CreativeGeneratorBlock;
import net.ndrei.teslacorelib.test.TeslaCoreUITestBlock;
import org.apache.logging.log4j.Logger;

@Mod(modid = TeslaCoreLib.MODID, version = TeslaCoreLib.VERSION, name = "Tesla Core Lib", dependencies = "after:tesla", useMetadata = true)
public class TeslaCoreLib
{
    public static final String MODID = "teslacorelib";
    public static final String VERSION = "@@VERSION@@";

    @Mod.Instance
    @SuppressWarnings("unused")
    public static TeslaCoreLib instance;

    public static Logger logger;

    public static ITeslaCorePackets network = new TeslaCorePackets(MODID);

    public static CreativeTabs creativeTab =  new CreativeTabs("tesla_core_lib") {
        @Override
        public ItemStack getIconItemStack()
        {
            return new ItemStack(TeslaCoreLib.wrench);
        }

        @Override
        public ItemStack getTabIconItem() { return this.getIconItemStack(); }
    };

    public static TeslaWrench wrench;
    public static TeslaBattery battery;
    public static BaseAddonItem baseAddon;
    public static MachineCaseItem machineCase;

    public static TeslaCoreUITestBlock testBlock;
    public static CreativeGeneratorBlock generatorBlock;

    public static GearWoodItem gearWood;
    public static GearStoneItem gearStone;
    public static GearIronItem gearIron;
    public static GearGoldItem gearGold;
    public static GearDiamondItem gearDiamond;

    public static EnergyUpgradeTier1 energyUpgradeTier1;
    public static EnergyUpgradeTier2 energyUpgradeTier2;
    public static SpeedUpgradeTier1 speedUpgradeTier1;
    public static SpeedUpgradeTier2 speedUpgradeTier2;

    @Mod.EventHandler
    @SuppressWarnings("unused")
    public void preInit(FMLPreInitializationEvent event){
        TeslaCoreLib.logger = event.getModLog();

        (TeslaCoreLib.gearWood = new GearWoodItem()).register();
        (TeslaCoreLib.gearStone = new GearStoneItem()).register();
        (TeslaCoreLib.gearIron = new GearIronItem()).register();
        (TeslaCoreLib.gearGold = new GearGoldItem()).register();
        (TeslaCoreLib.gearDiamond = new GearDiamondItem()).register();

        (TeslaCoreLib.wrench = new TeslaWrench()).register();
        (TeslaCoreLib.battery = new TeslaBattery()).register();
        (TeslaCoreLib.baseAddon = new BaseAddonItem()).register();
        (TeslaCoreLib.machineCase = new MachineCaseItem()).register();
        TeslaCoreCapabilities.register();

        (TeslaCoreLib.energyUpgradeTier1 = new EnergyUpgradeTier1()).register();
        (TeslaCoreLib.energyUpgradeTier2 = new EnergyUpgradeTier2()).register();
        (TeslaCoreLib.speedUpgradeTier1 = new SpeedUpgradeTier1()).register();
        (TeslaCoreLib.speedUpgradeTier2 = new SpeedUpgradeTier2()).register();

        TeslaCoreLib.testBlock = new TeslaCoreUITestBlock();
        TeslaCoreLib.testBlock.register();
        TeslaCoreLib.testBlock.setCreativeTab(TeslaCoreLib.creativeTab);

        TeslaCoreLib.generatorBlock = new CreativeGeneratorBlock();
        TeslaCoreLib.generatorBlock.register();
        TeslaCoreLib.generatorBlock.setCreativeTab(TeslaCoreLib.creativeTab);

        if (event.getSide() == Side.CLIENT) {
            TeslaCoreLib.gearWood.registerRenderer();
            TeslaCoreLib.gearStone.registerRenderer();
            TeslaCoreLib.gearIron.registerRenderer();
            TeslaCoreLib.gearGold.registerRenderer();
            TeslaCoreLib.gearDiamond.registerRenderer();

            TeslaCoreLib.wrench.registerRenderer();
            TeslaCoreLib.battery.registerRenderer();
            TeslaCoreLib.baseAddon.registerRenderer();
            TeslaCoreLib.machineCase.registerRenderer();

            TeslaCoreLib.energyUpgradeTier1.registerRenderer();
            TeslaCoreLib.energyUpgradeTier2.registerRenderer();
            TeslaCoreLib.speedUpgradeTier1.registerRenderer();
            TeslaCoreLib.speedUpgradeTier2.registerRenderer();

            TeslaCoreLib.testBlock.registerRenderer();
            TeslaCoreLib.generatorBlock.registerRenderer();
        }

        NetworkRegistry.INSTANCE.registerGuiHandler(TeslaCoreLib.instance, new TeslaCoreGuiProxy());
    }
}

File: src/main/kotlin/net/ndrei/teslacorelib/TeslaCoreLib.kt
package net.ndrei.teslacorelib

import net.minecraft.creativetab.CreativeTabs
import net.minecraft.item.ItemStack
import net.minecraftforge.fml.common.Loader
import net.minecraftforge.fml.common.Mod
import net.minecraftforge.fml.common.SidedProxy
import net.minecraftforge.fml.common.event.FMLConstructionEvent
import net.minecraftforge.fml.common.event.FMLInitializationEvent
import net.minecraftforge.fml.common.event.FMLPostInitializationEvent
import net.minecraftforge.fml.common.event.FMLPreInitializationEvent
import net.ndrei.teslacorelib.compatibility.RFPowerProxy
import net.ndrei.teslacorelib.config.IModConfigFlagsProvider
import net.ndrei.teslacorelib.config.TeslaCoreLibConfig
import net.ndrei.teslacorelib.config.TheGuiFactory
import net.ndrei.teslacorelib.energy.systems.MJSystem
import net.ndrei.teslacorelib.energy.systems.MekanismSystem
import net.ndrei.teslacorelib.energy.systems.TeslaSystem
import net.ndrei.teslacorelib.items.TeslaWrench
import net.ndrei.teslacorelib.netsync.ITeslaCorePackets
import net.ndrei.teslacorelib.netsync.TeslaCorePackets
import org.apache.logging.log4j.LogManager
import org.apache.logging.log4j.Logger

/**
 * Created by CF on 2017-06-28.
 */
@Mod(modid = MOD_ID, version = MOD_VERSION, name = MOD_NAME,
    acceptedMinecraftVersions = MOD_MC_VERSION,
    dependencies = "${MOD_DEPENDENCIES}after:${TeslaSystem.MODID};after:${RFPowerProxy.MODID};after:${MJSystem.MODID};after:${MekanismSystem.MODID}",
    useMetadata = true, modLanguage = "kotlin", modLanguageAdapter = "net.shadowfacts.forgelin.KotlinAdapter",
    guiFactory = TheGuiFactory.CLASS_NAME,
    certificateFingerprint = MOD_SIGN_FINGERPRINT
    )
object TeslaCoreLib : IModConfigFlagsProvider {
    @SidedProxy(clientSide = "net.ndrei.teslacorelib.ClientProxy", serverSide = "net.ndrei.teslacorelib.ServerProxy")
    lateinit var proxy: CommonProxy
    lateinit var logger: Logger

    val config = TeslaCoreLibConfig
    override val modConfigFlags = this.config

    // just as a 'proxy' for old code
    @Deprecated("Use the global constant MOD_ID instead.")
    const val MODID = MOD_ID // TODO: remove this

    val network: ITeslaCorePackets = TeslaCorePackets(MOD_ID)

    val creativeTab: CreativeTabs = object : CreativeTabs("tesla_core_lib") {
        override fun getTabIconItem() = ItemStack(TeslaWrench)
    }

    val isClientSide
        get() = net.minecraftforge.fml.common.FMLCommonHandler.instance().effectiveSide?.isClient ?: false

    @Mod.EventHandler
    fun construct(event: FMLConstructionEvent) {
        this.logger = LogManager.getLogger(Loader.instance().activeModContainer()!!.modId)

        this.proxy.construction(event)
    }

    @Mod.EventHandler
    fun preInit(event: FMLPreInitializationEvent) {
        this.config.init(event.suggestedConfigurationFile)
        RFPowerProxy.isRFAvailable = Loader.isModLoaded(RFPowerProxy.MODID)

        this.proxy.preInit(event)
    }

    @Mod.EventHandler
    fun init(event: FMLInitializationEvent) {
        this.proxy.init(event)
    }

    @Mod.EventHandler
    fun postInit(event: FMLPostInitializationEvent) {
        this.proxy.postInit(event)
    }
}
